openhtf/output/servers/station_server.py

```diff
@@ -0,0 +1,622 @@
+"""HTTP server serving station info and an Angular frontend / web GUI.
+
+This module is comparable to OpenHTF's output.web_gui module. Some differences:
+- Serves info about a single station only.
+- Designed to run in the same process as an HTF test, instead of using the
+      XML-RPC-based station API.
+"""
+
+import contextlib
+import hashlib
+import itertools
+import json
+import logging
+import os
+import re
+import socket
+import sys
+import threading
+import time
+import types
+
+import openhtf
+from openhtf.output.callbacks import mfg_inspector
+from openhtf.output.web_gui_server import pub_sub
+from openhtf.output.web_gui_server import web_gui_server
+from openhtf.util import conf
+from openhtf.util import data
+from openhtf.util import functions
+from openhtf.util import multicast
+from openhtf.util import timeouts
+import sockjs.tornado
+import tornado.web
+
+
+_LOG = logging.getLogger('chauffeurhtf.station_server')
+
+# Constants related to response times within the server.
+_CHECK_FOR_FINISHED_TEST_POLL_S = 0.5
+_DEFAULT_FRONTEND_THROTTLE_S = 0.15
+_WAIT_FOR_ANY_EVENT_POLL_S = 0.05
+_WAIT_FOR_EXECUTING_TEST_POLL_S = 0.1
+
+STATION_SERVER_TYPE = 'station'
+
+MIMETYPE_REVERSE_MAP = {v: k for k, v in mfg_inspector.MIMETYPE_MAP.iteritems()}
+TEST_STATUS_COMPLETED = 'COMPLETED'
+
+conf.declare('frontend_throttle_s', default_value=_DEFAULT_FRONTEND_THROTTLE_S,
+             description='Min wait between successive updates to the frontend.')
+conf.declare('station_server_port', default_value=0,
+             description='Port on which to serve the station server.')
+
+
+def _get_executing_test():
+  """Get the currently executing test and its state.
+
+  When this function returns, it is not guaranteed that the returned test is
+  still running. A consumer of this function that wants to access test.state is
+  exposed to a race condition in which test.state may become None at any time
+  due to the test finishing. To address this, in addition to returning the test
+  itself, this function returns the last known test state.
+
+  Returns:
+    test: The test that was executing when this function was called, or None.
+    test_state: The state of the executing test, or None.
+  """
+  tests = openhtf.Test.TEST_INSTANCES.values()
+
+  if not tests:
+    return None, None
+
+  if len(tests) > 1:
```

Comment:
TypeError: object of type 'generator' has no len()

Reply:
#727 Also fixed by your PR, although I've change it from `list(d.values())` to `list(six.itervalues(d))` to avoid creating the list twice in Python 2